odbc/src/api/odbc_uri.rs

```diff
@@ -0,0 +1,201 @@
+    // exist, it will only find the first.
+    fn remove_mandatory_attribute(&mut self, names: &[&str]) -> Result<&'a str> {
+        self.remove(names).ok_or_else(|| {
+            if names.len() == 1 {
+                ODBCError::InvalidUriFormat(format!(
+                    "{} is required for a valid Mongo ODBC Uri",
+                    names[0]
+                ))
+            } else {
+                ODBCError::InvalidUriFormat(format!(
+                    "One of {:?} is required for a valid Mongo ODBC Uri",
+                    names
+                ))
+            }
+        })
+    }
+
+    // remove_to_mongo_uri converts this ODBCUri to a mongo_uri String. It will
+    // remove all the attributes necessary to make a mongo_uri. This is destructive!
+    pub fn remove_to_mongo_uri(&mut self) -> Result<String> {
+        let user = self.remove_mandatory_attribute(USER)?;
+        let pwd = self.remove_mandatory_attribute(PWD)?;
+        // TODO SQL-990: Support the PORT attribute, right now the only way to specify PORT is as
+        // part of SERVER. If ports are specified in both SERVER and PORT and they do not match it
+        // should be an error (I think, check the spec if it says...).
+        let server = self.remove_mandatory_attribute(SERVER)?;
+        let ssl = self.remove(SSL);
+        // TODO SQL-990: we may wish to support more attributes as options.
+        // If we do, add more tests to cover them.
+        let ssl_string =
+            if ssl.is_some() && ssl.unwrap() != "0" && ssl.unwrap().to_lowercase() != "false" {
+                "?ssl=true"
+            } else {
+                ""
+            };
+        Ok(format!(
+            "mongodb://{}:{}@{}{}",
+            user, pwd, server, ssl_string
+        ))
+    }
+}
+
+mod unit {
+    // TODO SQL-990: Add more tests to cover the ODBC spec with regards to special characters.
+    #[test]
+    fn test_new() {
+        use super::*;
+        use crate::map;
+
+        assert!(ODBCUri::new("").is_err());
+        assert!(ODBCUri::new("Foo").is_err());
+        assert!(ODBCUri::new("driver=Foo;Bar").is_err());
+
+        // test one attribute
+        let expected = ODBCUri(map! {"driver".to_string() => "Foo"});
+        assert_eq!(expected, ODBCUri::new("Driver=Foo").unwrap());
+
+        // test two attributes
+        let expected = ODBCUri(map! {"driver".to_string() => "Foo", "server".to_string() => "bAr"});
+        assert_eq!(expected, ODBCUri::new("Driver=Foo;SERVER=bAr").unwrap());
+
+        // test that trailing ';' is not an issue
+        assert_eq!(expected, ODBCUri::new("Driver=Foo;SERVER=bAr;").unwrap());
+    }
+
+    #[test]
+    fn test_remove_to_mongo_uri() {
+        use super::*;
+        // test missing SERVER
+        assert!(ODBCUri::new("USER=foo;PWD=bar")
+            .unwrap()
+            .remove_to_mongo_uri()
+            .is_err());
```

Comment:
Should we be asserting specific errors here? I guess the specific thing we'd be looking for is the presence of the expected missing key in the error message. It is possible we edit `remove_to_mongo_uri` in the future such that `"USER=foo;PWD=bar" still encounters an error but a different error than "SERVER required...", which means we'll lose the coverage we think we currently have, right?

Reply:
I think the only thing that sort of works is checking the formatted error message. So I added that...